File: internal/authx/token.go
package authx

import (
	"context"
	"errors"
	"runtime/debug"
	"sync"
	"time"

	"github.com/roee-hersh/hallpass/internal/cache"
)

// Token is a bearer token with its expiry.
type Token struct {
	Value  string
	Expiry time.Time // zero means unknown; the source's TTL applies
}

// TokenSource obtains and caches a token, refreshing it shortly before it
// expires. Concurrent callers share one refresh.
type TokenSource struct {
	// Fetch obtains a fresh token.
	Fetch func(ctx context.Context) (Token, error)
	// Early is how long before expiry a refresh starts (default 5 min).
	Early time.Duration
	// DefaultTTL applies when a token has no known expiry (default 15 min).
	DefaultTTL time.Duration
	// Now is the clock (default time.Now).
	Now func() time.Time

	mu      sync.Mutex
	tok     Token
	exp     time.Time
	inflght *fetchCall
}

// fetchCall is one shared Fetch; done closes once tok and err are set.
type fetchCall struct {
	done chan struct{}
	tok  Token
	err  error
}

// defaultFetchTimeout bounds a Fetch whose caller's context has no deadline.
const defaultFetchTimeout = 30 * time.Second

// Get returns a valid token, fetching one if needed.
//
// The shared Fetch runs in its own goroutine on a context detached from the
// first caller's cancellation (cache.Detach): otherwise that caller going
// away would abort the fetch and hand every waiter a context.Canceled that
// is not theirs. Each caller stops waiting when its own ctx is done. A
// panic in Fetch becomes a *cache.PanicError for everyone waiting on it.
func (s *TokenSource) Get(ctx context.Context) (string, error) {
	if s.Fetch == nil {
		return "", errors.New("token source has no fetch function")
	}
	now := s.now()
	s.mu.Lock()
	if s.tok.Value != "" && now.Before(s.exp) {
		v := s.tok.Value
		s.mu.Unlock()
		return v, nil
	}
	fc := s.inflght
	if fc == nil {
		fc = &fetchCall{done: make(chan struct{})}
		s.inflght = fc
		fctx, cancel := cache.Detach(ctx, defaultFetchTimeout)
		go func() {
			defer cancel()
			s.fetch(fc, fctx, now)
		}()
	}
	s.mu.Unlock()
	select {
	case <-fc.done:
	case <-ctx.Done():
		return "", ctx.Err()
	}
	if fc.err != nil {
		return "", fc.err
	}
	return fc.tok.Value, nil
}

// fetch runs one Fetch for fc, then always clears the inflight call and
// closes fc.done, whether Fetch returned, panicked or called runtime.Goexit.
func (s *TokenSource) fetch(fc *fetchCall, ctx context.Context, now time.Time) {
	returned := false
	defer func() {
		if r := recover(); r != nil {
			fc.tok, fc.err = Token{}, &cache.PanicError{Value: r, Stack: debug.Stack()}
		} else if !returned {
			fc.tok, fc.err = Token{}, errors.New("token fetch exited without returning")
		}
		s.mu.Lock()
		s.inflght = nil
		if fc.err == nil {
			s.tok = fc.tok
			s.exp = s.expiryOf(fc.tok, now)
		}
		s.mu.Unlock()
		close(fc.done)
	}()
	fc.tok, fc.err = s.Fetch(ctx)
	returned = true
}

// Invalidate drops the cached token so the next Get fetches again. Call it
// when the upstream rejected the token.
func (s *TokenSource) Invalidate() {
	s.mu.Lock()
	s.tok = Token{}
	s.exp = time.Time{}
	s.mu.Unlock()
}

func (s *TokenSource) now() time.Time {
	if s.Now != nil {
		return s.Now()
	}
	return time.Now()
}

func (s *TokenSource) expiryOf(t Token, now time.Time) time.Time {
	early := s.Early
	if early == 0 {
		early = 5 * time.Minute
	}
	exp := t.Expiry
	if exp.IsZero() {
		ttl := s.DefaultTTL
		if ttl == 0 {
			ttl = 15 * time.Minute
		}
		exp = now.Add(ttl)
	}
	refreshAt := exp.Add(-early)
	// Very short-lived tokens: refresh at half life rather than immediately.
	if !refreshAt.After(now) {
		half := exp.Sub(now) / 2
		if half <= 0 {
			half = time.Second
		}
		refreshAt = now.Add(half)
	}
	return refreshAt
}

File: hallpass-py/src/hallpass/authx/token.py
"""A cached bearer token, refreshed shortly before it expires. Concurrent
callers share one refresh."""

from __future__ import annotations

import threading
import time
from collections.abc import Callable
from dataclasses import dataclass

from hallpass.core import evidence
from hallpass.core.cache import PanicError, detach, is_panic_type
from hallpass.core.context import Context

__all__ = ["DEFAULT_FETCH_TIMEOUT", "Token", "TokenSource"]

# Bounds a fetch whose caller's context has no deadline.
DEFAULT_FETCH_TIMEOUT = 30.0


@dataclass(frozen=True)
class Token:
    value: str = ""
    # Epoch seconds; None means unknown and the source's TTL applies.
    expiry: float | None = None


class _FetchCall:
    __slots__ = ("done", "err", "tok")

    def __init__(self) -> None:
        self.done = threading.Event()
        self.tok = Token()
        self.err: BaseException | None = None


class TokenSource:
    """Obtains and caches a token, refreshing it before it expires."""

    def __init__(
        self,
        fetch: Callable[[Context], Token] | None,
        early: float = 0.0,
        default_ttl: float = 0.0,
        now: Callable[[], float] | None = None,
    ) -> None:
        # Obtains a fresh token.
        self.fetch = fetch
        # How long before expiry a refresh starts (default 5 min).
        self.early = early
        # Applies when a token has no known expiry (default 15 min).
        self.default_ttl = default_ttl
        self.now = now
        self._lock = threading.Lock()
        self._tok = Token()
        self._exp = 0.0
        self._inflight: _FetchCall | None = None

    def _now(self) -> float:
        return self.now() if self.now is not None else time.time()

    def get(self, ctx: Context) -> str:
        """A valid token, fetching one if needed.

        The shared fetch runs in its own thread on a context detached from
        the first caller's cancellation. Each caller stops waiting when its
        own context ends. A crash in fetch is a PanicError for everyone.
        """
        if self.fetch is None:
            raise RuntimeError("token source has no fetch function")
        now = self._now()
        with self._lock:
            if self._tok.value and now < self._exp:
                return self._tok.value
            fc = self._inflight
            if fc is None:
                fc = _FetchCall()
                self._inflight = fc
                # A token exchange is not evidence for a decision and its
                # response carries the credential: never record it.
                fctx, cancel = detach(evidence.without_recorder(ctx), DEFAULT_FETCH_TIMEOUT)
                call = fc

                def run() -> None:
                    try:
                        self._fetch(call, fctx, now)
                    finally:
                        cancel()

                threading.Thread(target=run, name="hallpass-token", daemon=True).start()
        if not ctx.wait(fc.done):
            err = ctx.err()
            assert err is not None
            raise err
        if fc.err is not None:
            raise fc.err
        return fc.tok.value

    def _fetch(self, fc: _FetchCall, ctx: Context, now: float) -> None:
        assert self.fetch is not None
        try:
            try:
                fc.tok = self.fetch(ctx)
            except Exception as e:  # noqa: BLE001 - every failure reaches the waiters
                fc.tok = Token()
                fc.err = PanicError(e) if is_panic_type(e) else e
            except BaseException:  # noqa: BLE001 - Go's runtime.Goexit
                fc.tok = Token()
                fc.err = RuntimeError("token fetch exited without returning")
        finally:
            with self._lock:
                self._inflight = None
                if fc.err is None:
                    self._tok = fc.tok
                    self._exp = self._expiry_of(fc.tok, now)
            fc.done.set()

    def invalidate(self) -> None:
        """Drop the cached token so the next get fetches again. Call it
        when the upstream rejected the token."""
        with self._lock:
            self._tok = Token()
            self._exp = 0.0

    def _expiry_of(self, t: Token, now: float) -> float:
        early = self.early or 300.0
        exp = t.expiry
        if exp is None:
            exp = now + (self.default_ttl or 900.0)
        refresh_at = exp - early
        # Very short-lived tokens: refresh at half life, not immediately.
        if not refresh_at > now:
            half = (exp - now) / 2
            if half <= 0:
                half = 1.0
            refresh_at = now + half
        return refresh_at
